### PR TITLE
Core/Gameobject: Frozen trap (freezing arrow) target

### LINKED ISSUE #14238: [3.3.5] Frozen trap (freezing arrow) target
Core rev: https://github.com/TrinityCore/TrinityCore/commit/869bc4e0f7b07e6fe268b56b531fadde8cf15649

Steps to reproduce:
- Use freezing arrow near to an enemy and see how he gets frozen (it is okay, it's supposed to work like this)
- Wait and use freezing arrow again near to an enemy that is near the first frozen enemy
- The target that was frozen first is the enemy that gets frozen instead of the enemy nearest to the trap

Video proof so you can see it better: https://www.youtube.com/watch?v=kFDJWtHy3FI

``` diff
diff --git a/src/server/game/Entities/GameObject/GameObject.cpp b/src/server/game/Entities/GameObject/GameObject.cpp
index 6ca5ab4..861e041 100644
--- a/src/server/game/Entities/GameObject/GameObject.cpp
+++ b/src/server/game/Entities/GameObject/GameObject.cpp
@@ -475,11 +475,18 @@ void GameObject::Update(uint32 diff)
                     if (Unit* owner = GetOwner())
                     {
                         // Hunter trap: Search units which are unfriendly to the trap's owner
-                        Trinity::AnyUnfriendlyNoTotemUnitInObjectRangeCheck checker(this, owner, radius);
-                        Trinity::UnitSearcher<Trinity::AnyUnfriendlyNoTotemUnitInObjectRangeCheck> searcher(this, target, checker);
-                        VisitNearbyGridObject(radius, searcher);
+                        UnitList targets;
+                        Trinity::AnyUnfriendlyNoTotemUnitInObjectRangeCheck u_check(this, owner, radius);
+                        Trinity::UnitListSearcher<Trinity::AnyUnfriendlyNoTotemUnitInObjectRangeCheck> searcher(this, targets, u_check);
+                        VisitNearbyObject(radius, searcher);
                         if (!target)
                             VisitNearbyWorldObject(radius, searcher);
+
+                        if (!targets.empty())
+                        {
+                            targets.sort(Trinity::ObjectDistanceOrderPred(this));
+                            target = targets.front();
+                        }
                     }
                     else
                     {
```
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8876527-3-3-5-frozen-trap-freezing-arrow-target?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Aokromes**
**Target branch(es)**: 335/6x

**Issues addressed**: Closes #14238

**Tests performed**: (Does it build, tested in-game, etc)

By robinsch

### CODE REVIEW

**Comment by Shauren** [comment]
Good idea, bad way to use grid searchers. Need to implement a new NearestUnfriendlyNoTotemUnitInObjectRangeCheck and use it together with UnitLastSearcher so that you get one unit in a single step instead of a list that has to be sorted later

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 1]
```diff
@@ -502,11 +502,18 @@ void GameObject::Update(uint32 diff)
                     if (Unit* owner = GetOwner())
                     {
                         // Hunter trap: Search units which are unfriendly to the trap's owner
-                        Trinity::AnyUnfriendlyNoTotemUnitInObjectRangeCheck checker(this, owner, radius);
-                        Trinity::UnitSearcher<Trinity::AnyUnfriendlyNoTotemUnitInObjectRangeCheck> searcher(this, target, checker);
-                        VisitNearbyGridObject(radius, searcher);
+                        UnitList targets;
+                        Trinity::AnyUnfriendlyNoTotemUnitInObjectRangeCheck u_check(this, owner, radius);
```
Comment (jackpoz) [inline src/server/game/Entities/GameObject/GameObject.cpp]: "checker" was bad as name ?

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 2]
```diff
@@ -502,11 +502,18 @@ void GameObject::Update(uint32 diff)
...
+                        Trinity::UnitListSearcher<Trinity::AnyUnfriendlyNoTotemUnitInObjectRangeCheck> searcher(this, targets, u_check);
+                        VisitNearbyObject(radius, searcher);
                         if (!target)
```
Comment (jackpoz) [inline src/server/game/Entities/GameObject/GameObject.cpp]: is this supposed to still check "target" ?

File: src/server/game/Entities/GameObject/GameObject.cpp (hunk 1 above)
Comment (Aokromes) [inline src/server/game/Entities/GameObject/GameObject.cpp]: @robinsch 
https://github.com/TrinityCore/TrinityCore/pull/17517#discussion_r69759509
https://github.com/TrinityCore/TrinityCore/pull/17517#discussion_r69759675

**Comment by Aokromes** [comment]
@robinsch 
#17517 (comment)
#17517 (comment)

**Pull request closed without merging**